include/mfmg/dealii/lanczos.templates.hpp

```diff
@@ -0,0 +1,342 @@
+  evals.resize(n_eigenvectors);
+  evecs.resize(n_eigenvectors);
+
+  return std::make_tuple(evals, evecs);
+}
+
+/// \brief Lanczos solver: perform Lanczos solve
+template <typename OperatorType, typename VectorType>
+template <typename FullOperatorType>
+std::tuple<std::vector<double>, std::vector<VectorType>>
+Lanczos<OperatorType, VectorType>::details_solve_lanczos(
+    FullOperatorType const &op, const int num_requested,
+    boost::property_tree::ptree const &params, VectorType const &initial_guess)
+{
+  const int maxit = params.get<int>("max_iterations");
+  const double tol = params.get<double>("tolerance");
+  const int percent_overshoot = params.get<int>("percent_overshoot", 0);
+
+  ASSERT(0 <= percent_overshoot && percent_overshoot < 100,
+         "Lanczos overshoot percentage should be in [0, 100)");
+  ASSERT(tol >= 0., "Lanczos tolerance must be non-negative");
+  ASSERT(maxit >= num_requested, "Lanczos max iterations is too small to "
+                                 "produce required number of eigenvectors.");
+
+  const int n = op.n();
+
+  std::vector<double> evals;
+  std::vector<VectorType> evecs;
+
+  // Initializations; first Lanczos vector.
+  double alpha = 0;
+  double beta = initial_guess.l2_norm();
+
+  std::vector<VectorType> lanc_vectors; // Lanczos vectors
+
+  // Create first Lanczos vector if necessary
+  if (lanc_vectors.size() < 1)
+    lanc_vectors.push_back(initial_guess);
+
+  std::vector<double> t_maindiag;
+  std::vector<double> t_offdiag;
+
+  std::vector<double>
+      evecs_tridiag; // eigenvectors of tridiagonal matrix, stored in flat array
+
+  // Lanczos iteration loop
+  int it = 1;
+  for (int it_prev_check = 0; it <= maxit; ++it)
+  {
+    // Normalize lanczos vector
+    assert(beta != 0); // TODO: set up better check for near-zero
+    lanc_vectors[it - 1] /= beta;
+
+    if (lanc_vectors.size() < static_cast<size_t>(it + 1))
+    {
+      // Add new Lanczos vector
+      lanc_vectors.push_back(VectorType(n));
+    }
+
+    // Apply operator.
+    op.vmult(lanc_vectors[it], lanc_vectors[it - 1]);
+
+    // Compute, apply, save Lanczos coefficients
+    if (it != 1)
+    {
+      lanc_vectors[it].add(-beta, lanc_vectors[it - 2]);
+      t_offdiag.push_back(beta);
+    }
+
+    alpha = lanc_vectors[it - 1] * lanc_vectors[it]; // = tridiag_{it,it}
+
+    t_maindiag.push_back(alpha);
+
+    lanc_vectors[it].add(-alpha, lanc_vectors[it - 1]);
+
+    beta = lanc_vectors[it].l2_norm(); // = tridiag_{it+1,it}
+
+    // Check convergence if requested
+    // NOTE: an alternative here for p > 0 is
+    // int((100./p)*ln(it)) > int((100./p)*ln(it-1))
+    bool const first_iteration = (it == 1);
+    bool const max_iterations_reached = (it == maxit);
+    bool const percent_overshoot_exceeded =
+        (100 * (it - it_prev_check) > percent_overshoot * it_prev_check);
+    if (first_iteration || max_iterations_reached || percent_overshoot_exceeded)
+    {
+      const int dim_hessenberg = it;
```

Comment:
Declare it unsigned and the hell with cast in the assertion below.

Reply:
It would fix the assert, but would cascade later for comparisons with `num_requested` and such. Just leaving this alone.